src/Featurizers/Components/NormUpdaters.h

```diff
@@ -0,0 +1,413 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+#include "../../Traits.h"
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+namespace TypeSelector {
+// max norm type selector is created for selecting different types for max norm based on input type
+// when input type is integer, uint64 is chosen for performance
+// when input type is numeric, long double is chosen
+template <typename T>
+struct MaxNormTypeSelector {
+    using type = std::uint64_t;
+};
+template <>
+struct MaxNormTypeSelector<std::float_t> {
+    using type = long double;
+};
+
+template <>
+struct MaxNormTypeSelector<std::double_t> {
+    using type = long double;
+};
+
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+
+namespace Details{
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         L1NormUpdater
+///  \brief         L1NormUpdater updates l1 norm
+///                 l1 norm is the sum of elements absolute value
+///
+template <typename T>
+class L1NormUpdater {
+public:
+
+    /////////////////////////////////////////////////////////////////////////
+    ///  \class         L1NormResult
+    ///  \brief         A structure which contains the l1 norm
+    ///
+    struct L1NormResult {
+    public:
+        // ----------------------------------------------------------------------
+        // |
+        // |  Public Data
+        // |
+        // ----------------------------------------------------------------------
+        long double                                         const L1_norm;
+
+        // ----------------------------------------------------------------------
+        // |
+        // |  Public Methods
+        // |
+        // ----------------------------------------------------------------------
+        L1NormResult(long double l1_norm);
+        ~L1NormResult(void) = default;
+
+        FEATURIZER_MOVE_CONSTRUCTOR_ONLY(L1NormResult);
+    };
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                         = T;
+
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    L1NormUpdater(void);
+
+    void update_l1(InputType input);
+
+    L1NormResult commit_l1(void);

```

Comment:
This can probably be `commit`

Reply:
Now that I think about this, all of these updaters return a single value. Should this return a long double rather than a struct?